### DISCUSSION TITLE
Generated prisma types are missing relationships in their exported type.

### DISCUSSION THREAD

**Discussion by deathlynocturnal** [prisma/prisma#10928] [Q&A] [answered]
### Bug description

So I am really not sure if this is a bug, my mistake or some really stupid feature.
I started using prisma in force a few days ago and so far, I love it, but I am struggling to understand how to use the generated types from `@prisma/client` for example:

I have a model called Faction, this faction creates a one to many relationship with users (noted as `owner`), so each faction is owned by a single user, but a user can own many factions.

Oddly, if I export `Faction` from the generated type by the `@prisma/client` it seems to be that if I was to do:
`faction.owner` will return as it does not exist, but in the model, it does exist, and if I was to do `include: { owner: true }` it's fine, but how do I type this for TypeScript so I can pass data to and from different places/functions/methods, etc.

For example I am working on a helper where I am running the same function (gets the faction through a different relationship) but I can't actually type it, because if I set the return as `Faction` it comes back with no `owner` attached therefore I just get `owner does not exist`.

### How to reproduce

1. Use the model below, I cut out a lot of other stuff that isn't necessary, as this model is a lot bigger.
2. Generate the client, set a type as the `Faction` type exported from `@prisma/client`.
3. Observe how `owner` is absent from that type.

### Expected behavior

I would expect, that there is some functionality that allows me to use the types generated by prisma that will contain the owner as well as the faction and should have some kind of logic to allow this?

### Prisma information

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["mongoDb"]
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

enum FactionChannelType {
  TEXT
  VOICE
  NEWS
  STAGE
}

// Core Models
model User {
  id            String           @id @default(dbgenerated()) @map("_id") @db.ObjectId
  discordId     String           @unique
  username      String
  discriminator String
  ownedFactions Faction[]        @relation(name: "OwnedFactions")
}

model Faction {
  id          String           @id @default(dbgenerated()) @map("_id") @db.ObjectId
  name        String
  roleId      String
  categoryId  String
  owner       User             @relation(name: "OwnedFactions", fields: [ownerId], references: [id])
  ownerId     String
}
```

### Environment & setup

- OS: Ubuntu 20.04 focal
- Database: MongoDB
- Node: 16.13.0
- IDE: VSCode (Remote - SSH)

### Prisma Version

```
prisma                  : 3.7.0
@prisma/client          : 3.7.0
Current platform        : debian-openssl-1.1.x
Query Engine (Node-API) : libquery-engine 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/@prisma/engines/libquery_engine-debian-openssl-1.1.x.so.node)
Migration Engine        : migration-engine-cli 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine    : introspection-core 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary           : prisma-fmt 8746e055198f517658c08a0c426c7eec87f5a85f (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Default Engines Hash    : 8746e055198f517658c08a0c426c7eec87f5a85f
Studio                  : 0.445.0
Preview Features        : mongoDb
```

**Comment by austincrim** [comment, marked as answer]
Hey @dannysmc95 👋 !

Since Prisma queries do not include relations by default (you have to use the `include` option), the generated types do not include them either. You can create the type you are looking for using one of our built-in utility types, though.

```ts
import { Prisma } from '@prisma/client'
type FactionWithOwner = Prisma.FactionGetPayload<{
  include: { owner: true }
}>
```

Here we use the `FactionGetPayload` type and pass in some query options as the generic parameter to get the desired type.

Let me know if you have any questions about this!

**Reply by deathlynocturnal** [reply]
Amazing, thank you for that!!

**Reply by bel-mont** [reply]
Is there absolutely no other way around this to get the generator to do it? I'm trying to move away from sequelize in a fairly big database (150ish tables), and doing this for all necessary cases would take a long time.

**Reply by dthuering** [reply]
Hello,
how to use this type in a query now?
I cant seem to get any relationship data returned. I am using MongoDB.
Best,
Danny

**Reply by dthuering** [reply]
The issue was that I did not run the `npx db push` after adding the relation which created some index.

**Reply by renatoaraujoc** [reply]
Hello @austincrim,

Thanks for your input and that's the real answer, but I'm facing Intellisense issues (though the code works!). Using the latest version of Intellij IDEA right now.

Intellisense shows this:
![Captura de Tela 2022-03-27 às 14 22 12](https://user-images.githubusercontent.com/358128/160293010-ec93341b-f95a-4c78-9704-fb79c9545678.png)

It fails if I input a relationship which does not exist, which is correct:
![Captura de Tela 2022-03-27 às 14 23 56](https://user-images.githubusercontent.com/358128/160293062-202f6266-454c-464c-8b65-0946d8f6ae47.png)

This is the final result (I hope it's right)
![Captura de Tela 2022-03-27 às 14 24 37](https://user-images.githubusercontent.com/358128/160293093-7ce5892e-fa29-406b-8183-6df2c1c09313.png)

Right now the only issue I'm having is intellisense, does it make any sense?

Thank you!

Renato

**Reply by wanjas** [reply]
This [answer](https://github.com/prisma/prisma/discussions/10928#discussioncomment-4791243) [thw0rted](https://github.com/thw0rted)  mentioned doc's article [Operating against partial structures of your model types](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types) but on top of that there is a blog article [How TypeScript 4.9 `satisfies` Your Prisma Workflows](https://www.prisma.io/blog/satisfies-operator-ur8ys8ccq7zb) on using `satisfy` to make life even easier when you want to create such types.

**Reply by sjoerdvanBommel** [reply]
For those trying to improve their TypeScript experience a little further: you could also override the generated type from `@prisma/client` or export it as a separate type like this, where the relationships are optional.

```
import {
  Prisma,
  Question as PrismaQuestion,
} from '@prisma/client'

...

export type Question = PrismaQuestion &
    Partial<Prisma.QuestionGetPayload<{ include: { answers: true; question_type: true } }>>
```

Results in type:

```
{
    id: number;
    question_text: string;
    difficulty_level: number;
    question_type_id: number;
    answers?: {
        ...
    }[] | undefined;
    question_type?: {
        ...;
    } | undefined;
}
```

**Reply by pablojsx** [reply]
thank you!

**Reply by adityapadekar** [reply]
@austincrim Is there any way to have partial owner object?

Rn I'm facing a situation in which I wish to have only two fields from the referenced object lets say username and discordId from the owner.

I want to do this to improve performance and we don't need the full object which is quite massive.

**Reply by thw0rted** [reply]
Sounds like you're asking for https://www.prisma.io/docs/orm/prisma-client/queries/select-fields ?

**Reply by adityapadekar** [reply]
@thw0rted No, not that 

My query is as follows ( this is a server action )

```javascript
const referralCodes = await db.referralCode.findMany({
    select: {
      id: true,
      code: true,
      status: true,
      createdBy: true,
      createdAt: true,
      user: {
        select: {
          id: true,
          name: true,
          email: true,
        },
      },
    },
    orderBy: {
      createdAt: "desc",
    },
  });
```

### Now the issue being the generate type by prima does not include the user as it is referenced as follows

```prisma
model ReferralCode {
  id     String             @id @default(auto()) @map("_id") @db.ObjectId
  code   String             @unique
  status ReferralCodeStatus @default(ACTIVE)

  createdBy String   @map("user") @db.ObjectId
  createdAt DateTime @default(now())

  user User @relation(fields: [createdBy], references: [id], onDelete: Cascade)
}
```

### The ReferralCode type imported from @prisma/client includes id, code, status, createdBy and createdAt but no user in it

Now I managed to solve the issue using the code snippet below 
```javascript
import { Prisma, ReferralCode, User } from "@prisma/client";

const adminPersonalData = Prisma.validator<Prisma.ReferralCodeDefaultArgs>()({
  include: { user: { select: { id: true, email: true, name: true } } },
});

export type ReferralCodeWithAdminPersonalData = Prisma.ReferralCodeGetPayload<typeof adminPersonalData>;
```

here I am selecting certain fields user but one can retrieve whole user by using the following
```javascript
const adminPersonalData = Prisma.validator<Prisma.ReferralCodeDefaultArgs>()({
  include: { user: true } },
});
```


**NOW what I wished to implement it a dynamic way to include the user object where not all fields of the user are required but I also don't want to explicitly mention which will be included in a particular response like done in the 1st example**

**Reply by thw0rted** [reply]
It looks like you already understand how to use the GetPayload helper type to include certain fields from your User model in the results, and also how to include all the user fields.  What's the problem, then?  If you want the types to include all fields, `user: true`, if not, `select: {....}`.  If the problem is that the fields aren't "optional", and you want to be forced to check if they exist before using them, you could just pass them through a `Partial<...>` helper.

**Reply by adityapadekar** [reply]
@thw0rted 
To understand what I want let's take one example 

**Case 1:** Fetch Referral Code documents with user's name and email.
For this case I need only these two field.

**Case 2:** Fetch Referral Code documents with all the user's data 

**Case 3:** Fetch Referral Code document with user's name, email, credit and status fields. 

Now to do this I need to create a new type for each scenario and some of the fields are not optional in the user document

I attempted to use `Partial<>` to achieve this but encountered issues. Here's what I tried:

```javascript 
const adminPersonalData = Prisma.validator<Prisma.ReferralCodeDefaultArgs>()({
  include: { user: true } },
});

export type ReferralCodeWithAdminPersonalData = Prisma.ReferralCodeGetPayload<Partial<typeof adminPersonalData>>;
```
This results in the following type for  ReferralCodeWithAdminPersonalData
```typescript
type exp = {
    id: string;
    code: string;
    status: $Enums.ReferralCodeStatus;
    createdBy: string;
    createdAt: Date;
}
```

However, this approach didn't work as expected. I understand I could create separate types for each scenario, but that would result in redundant code. Also, fetching the whole object isn't efficient.

Is there a better way to achieve this? Am I missing something in my approach? Any suggestions or guidance would be greatly appreciated. Thank you!

**Reply by thw0rted** [reply]
I found it helpful to stop writing so many explicit types everywhere. If you let TS infer return types you can sometimes avoid having to write a lot of variations, without losing strong typing.

**Reply by adityapadekar** [reply]
@thw0rted I'm a bit confused as to what you mean by that.. can you provide a example?

**Reply by thw0rted** [reply]
It's more general advice.  I used to write explicit return types on lookup methods:

```
type UserWithMinimalPosts = Prisma.UserGetPayload<{include: {post: select: typeof MINIMAL_POST_SELECT}}}>;
function getUser(...): UserWithMinimalPosts { ... }
```

I realized that the return type is redundant as long as it matches the output of my `prisma.user.findUnique(...)` call, and actively harmful if it isn't exactly the same.  Variable typing (`const user: UserWithMinimalPosts = getUser(...)`) is the same.

Instead, I focus on typing function arguments.  If a function only actually uses the username, email, and post contents, I can write the argument type as

```
type UserWithMinimalPosts = Prisma.UserGetPayload<{include: {post: {select: typeof MINIMAL_POST_SELECT}}}>;
function renderList(user: Pick<UserWithMinimalPosts, 'username' | 'email' | 'post'>) { ... }
```

This way, I can pass the response from a Prisma lookup that only returns a few user fields, or one that returns their full profile, as long as it includes at least the basic fields from the related Post model, but the typechecker will complain if I try to pass a response from a query with no `include: { post: ... }` at all.

If this isn't helpful, maybe you could have an example of how your types are being used?

**Reply by adityapadekar** [reply]
@thw0rted okay!... I'll look init, I'm new to prisma and also typescript to some extent, so I don't necessarily get everything you said but I'll run few experiment to understand this.

Thanks for everything.

**Comment by magrinj** [comment]
If you do a fallback call of your prisma client functions, you can just type it correctly like below and you should have the same behaviour as if you directly call the prisma client:

```typescript
 async findMany<T>(args: Prisma.SelectSubset<T, Prisma.FactionFindManyArgs>) {
    return this.prismaClient.faction.findMany<T>(args);
  }
```

**Reply by jumantet** [reply]
Great, that works for me ! Many thanks !

**Reply by JefferyHus** [reply]
As of `4.16.2` this doesn't work :/

**Reply by ErnestoSalazar** [reply]
Agree with @JefferyHus, I had to make some adjustments to make it work in a newer version.

```ts
async findMany<T extends Prisma.FactionFindManyArgs>(args: Prisma.SelectSubset<T, Prisma.FactionFindManyArgs>) {
    return this.prismaClient.faction.findMany<T>(args);
  }
```

Here's another example with types in a separate file:

__types.ts__
```ts
export type UpdateUserParams<T> = Prisma.SelectSubset<T, Prisma.UserUpdateArgs>;
```

__updateUser.ts__
```ts
export const updateUser = async <T extends Prisma.UserUpdateArgs>(
  args: UpdatePlayerDbParams<T>,
) => {
  return prisma.user.update<T>(args);
}
```

This is the only solution that has worked for me so far, the function should still maintain the same behavior as if you call Prisma client directly just as magrinj mentions. And this will prevent creating and maintaining custom types.

I hope someone finds this useful.

**Comment by George3d6** [comment]
Is there really no better option to get proper typing here? This seems like a very-very common usecase and it adds tremendous complexity.

**Reply by Honatas** [reply]
I agree. It doesn't make sense to have to create my own types just to get my code that uses these relationships to transpile. The typecasts also look really ugly.

**Reply by gregg-cbs** [reply]
There 100% should be a way to generate types with relationships. Stringing this together manually and then having to keep the models and manually created relation types in sync is becoming a nightmare. Never mind when you sometimes import types from prisma folder and sometimes import them from your create types folder it becomes confusing - especially when intellisense recommends prismas non relation types first.

**Reply by conterio** [reply]
it shouldn't be this hard, let me just dev

**Reply by thw0rted** [reply]
It isn't "hard"; did you read the accepted answer?  I had to spend some time with it to get a handle on how to use them, but the `Prisma.{model}GetPayload` types are exactly what everybody here seems to be asking for.  I think the docs do a pretty good job of explaining how to use them now.

**Reply by swalker326** [reply]
Agreed I don't understand the outrage here. You can generate the type however you want. Prisma would have to generate every combination of relation and give it some default name. This gives the dev full control of the naming and included relations.

**Comment by KeentGG** [comment]
We should atleast add `GetPayload` to the Utility Types documentation, there's literally only one documented which is `validator`.

`GetPayload` is an essential tool for strict type-safety projects, especially i'll be using this with NestJS.

**Comment by SawkaDev** [comment]
+1 I think this should be a feature request or at least add to the documentation how to handle this use case for Typescript. Industry is moving towards Typescript, having this out of the box would be nice.

**Comment by thw0rted** [comment]
A similar discussion included a link to [this docs page](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types) about advanced type usage, and I thought it might be helpful here as well.

**Reply by kilroyjones** [reply]
How do I even get validator to work? I keep getting that it's not defined on PrismaClient.

**Reply by ngharry** [reply]
Thanks heaps, it works like a charm for me

**Reply by bursteways** [reply]
Nice one, thanks. @thw0rted

**Comment by thetminko** [comment]
This is one of the pain that I am having also. When there are different variations, the codebase gets bloated with so many custom types. It would be great if generator can provide a type with optional relationships, so the developer can decide to use that or create custom types if want to be specific.

**Comment by gregg-cbs** [comment]
You can use Zod Prisma Types plugin for this. It generates zod schemas which you can ignore but it also generates types with relations .

**Reply by thetminko** [reply]
thanks for the suggestion!

**Comment by JefferyHus** [comment]
Frustrating, I tried every solution posted here and it just doesn't work. It is just one more reason to really just drop Prisma and go back to using a better ORM :)

**Reply by JefferyHus** [reply]
The only way I was able to go around this is by doing this

```typescript
Prisma.UserGetPayload<{
    include: Prisma.UserInclude<any>;
}>
```

I tried so many thing instead of `any` but nothing worked as expected

**Comment by nla-asia** [comment]
`
export type Article = Prisma.ArticleGetPayload<{
  include: { author: true },
}>;
`
I can include the relationship like this include all the fields from 'author' . Is there any way to define which fields from the relationship to include ?

**Comment by david1542** [comment]
I also need this type of functionality. This seems very basic and needed. Isn't there a way to add a flag to `prisma generate` that will also generate the relations themselves?

**Comment by danthedaniel** [comment]
This is what I do:

```typescript
import { Prisma } from "@prisma/client";
import { Types, OperationPayload } from "@prisma/client/runtime/library";

type PayloadToModel<T> =
    T extends OperationPayload ? Types.Result.DefaultSelection<T> :
    T extends OperationPayload[] ? Types.Result.DefaultSelection<T[number]>[] :
    never;

export type WithRelation<
    T extends keyof Prisma.TypeMap["model"],
    R extends keyof Prisma.TypeMap["model"][T]["payload"]["objects"]
> =
    PayloadToModel<Prisma.TypeMap["model"][T]["payload"]> &
    {
      [K in R]: PayloadToModel<Prisma.TypeMap["model"][T]["payload"]["objects"][K]>
    };
```

And then reference it as:

```typescript
import type { WithRelation } from "@helpers";

function foo(model: WithRelation<"User", "comments" | "posts">) { ... }
```

I wrote out a bunch of types to make it so that it takes in `User` instead of `"User"` as a string, but it was way more complicated and in the end, unnecessary.

**Edit**:

Updated for Prisma 5.16

**Edit 2**:

Updated for Prisma 6.16:

* `./generated` is the output path as specified in `prisma.schema`
* It's been a while since I touched Prisma, so the nullability code added to `PayloadToModel` seems like one of those "why did this work before?" things, which is a red flag. But... it works!

```typescript
import type { OperationPayload, Types } from "@prisma/client/runtime/library";
import type { Prisma } from "./generated";

type CopyNullability<T, U> = null extends T ? U | null : U;
type PayloadToModel<T, U = NonNullable<T>> =
  U extends OperationPayload ? CopyNullability<T, Types.Result.DefaultSelection<U>> :
  T extends OperationPayload[] ? Types.Result.DefaultSelection<T[number]>[]
  : never;

type ModelNames = keyof Prisma.TypeMap["model"];
type RelationNames<T extends ModelNames> =
  keyof Prisma.TypeMap["model"][T]["payload"]["objects"];

export type WithRelation<T extends ModelNames, R extends RelationNames<T>> =
  PayloadToModel<Prisma.TypeMap["model"][T]["payload"]> &
  {
    [K in R]: PayloadToModel<Prisma.TypeMap["model"][T]["payload"]["objects"][R]>
  };
```

**Comment by AbdelrahmanSadat** [comment]
Came here through Google looking after facing the same issue. The first answer worked, but I imagine there can be a better-scaling solution, as I imagine this can get tedious quickly in larger projects.

**Comment by travelerr** [comment]
I’m using the marked answer for now but I agree, this is a super common use case and a better solution would be very useful. +1

**Comment by swalker326** [comment]
Since this is the first google search result, I'll add my comment here. 

from the [prisma docs](https://www.prisma.io/docs/orm/prisma-client/type-safety/operating-against-partial-structures-of-model-types#solution)

```ts 
const videoWithVotes = Prisma.validator<Prisma.VideoDefaultArgs>()({
  include: { votes: true }
});
type VideosWithVotes = Prisma.VideoGetPayload<typeof videoWithVotes>;
```

**Reply by ZeldOcarina** [reply]
Yes this should be highlighted as the native way to go imo

**Comment by ericfel93** [comment]
i would expect prisma to generate `VideoWithRelations` type where all relationships are optional alongside with `Video` type

or maybe even have it on `Video` and not in another class, and not only the scalars

**Reply by swalker326** [reply]
I think making the type optional has interesting consequences with trying to access keys. Now typescript will want them checked for exsitence before getting any data out of them. Also the relations could have relations, would these also be optional attributes? 

Could see a chain of 
```ts
const relation = root.relation?.relations_relation?.relations_relations_relation
if (!relation) {
//handle this
}
```

**Reply by ericfel93** [reply]
yep, looking good IMO

**Reply by thw0rted** [reply]
In case you haven't looked at the accepted answer yet, what you're suggesting probably isn't needed.  The issue is that if you query an object without an `include` for the relationship, you won't get the related record.  You don't need the relationship to be optional in that case, you absolute are not going to have it.  You just have to trust the return types for the query methods.

As an aside, I just started writing new Prisma-using TS code after a long time maintaining an old vanilla-JS codebase, where there was very little type information (`/** @type ... */` etc).  It took me a while to figure out why I had so much trouble with related records.  The root issue was that we were explicitly typing a bunch of stuff using the shorthand / default Model types, like `import { User } from '@prisma/client'`.  It's too easy to assume that the type with the simplest name is probably the one you want to use everywhere.  As soon as I started stripping away explicit `User` return types and just letting TS carry the correct (very complex) return types around, everything was so much simpler.  That, combined with learning to use e.g. `UserGetPayload`, made a huge difference.

**Reply by ericfel93** [reply]
my issue with the accepted solution is that I'm going to have many models
in the above example you gonna need to have one `root` object that all of the inner includes are optional, than you going to have `relation` that as its all includes and on and on
and regards the fact that even if you didn't included anything in the query you are going to have an optional relationship, IMO it is good. working with SqlAlchemy I got optional and it worked amazing

**Reply by ericfel93** [reply]
it can be a parameter passed to `prisma-client-js` whether to create optional relationship fields or not

**Comment by nawangwe** [comment]
Found this library which might be of some help for now: [https://github.com/Kalissaac/prisma-typegen](https://github.com/Kalissaac/prisma-typegen)

**Reply by wanjas** [reply]
This library or optional relations in general looks to me as an anti-patter. Relations in Prisma are fundamentally dynamic. 

So using Prisma's built-in tools or using getters is a better practice.

## Prisma's way
```typescript
const userEntityInclude = {
  include: {
    posts: true,
    profile: true,
    etc: true,
  },
} satisfies Prisma.UserDefaultArgs;

export type UserEntity = Prisma.UserGetPayload<
  typeof userEntityInclude
>;
```

## Getter's return type

```typescript
import type { AsyncReturnType } from 'type-fest';

async function getUserEntity(id: string) {
  return prisma.user.findUnique({
    where: { id },
    include: {
      posts: true,
      profile: true,
      etc: true,
    },
  });
}

type UserEntity = AsyncReturnType<typeof getUserEntity>;
```

P.S. Both these ways provide you type-safety for even very complex includes/selects like the following example

```typescript
async function getUserEntity(id: string) {
  return prisma.user.findUnique({
    where: { id },
    include: {
      posts: {
        include: {
          tags: true,
          categories: {
            select: {
              id: true,
              include: {
                translations: true,
              },
            },
          },
        },
      },
      profile: {
        select: {
          name: true,
          avatar: true,
        },
      },
      etc: true,
    },
  });
}


```

**Reply by nawangwe** [reply]
Thank you for that insight, I understand your point completely. I guess I would still consider that library if the number one item in my project rubric was rapid prototyping and not having to worry about updating my type declaration on my getter. But your method is what I'm gonna go with.

**Reply by thw0rted** [reply]
You don't need the 3rd party utility library, it's [baked into Prisma](https://www.prisma.io/docs/orm/prisma-client/type-safety/operating-against-partial-structures-of-model-types#problem-getting-access-to-the-return-type-of-a-function):

```ts
import { Prisma } from '@prisma/client'

type UsersWithPosts = Prisma.PromiseReturnType<typeof getUsersWithPosts>
```

**Comment by mmaazahmed-diya** [comment]
why aren't they included in by default? my work around, invocles a deepy nested type decleration. surely there must be a better way?
type test= Prisma.orderGetPayload<{include:{order_items:{include:{partnerLocation:{include:{city:true}}}}}}>

**Comment by HaulinOats** [comment]
You can always just make your own type this way too by pulling in default types: 
```
User[] & { posts: Post[]}
```

**Comment by Henson-Kudi** [comment]
This does not work for me in the case of nested include statements as the return type will only return the fields added in the include.
example:
model User{
...
roles: Role[]
}
model Role{
...
permissions: Permission[]
}

model Permission{
...}



const userWithRoles = Prisma.validator<Prisma.UserDefaultArgs>()({
  include: {
    roles: {
      include: {
        permissions: true
      }
    },
  },
})

// 3: This type will include a user and all their posts
export type UserWithRoles = Prisma.UserGetPayload<typeof userWithRoles>

The above type does not include the properties of roles except permissions which is explicitly added in the include statement.

In order to solve this, I had to explicitly select the fields for roles which is not the best approach I guess?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by Destroy666x** [comment]
Weird decision to not include such types in Prisma generated client by default as e.g. `UserWithRelations`, `FullUser` or whatever.

What would be the best way currently to autogenerate all the possible relation types? I've seen Zod + plugin mentioned, anything else?

**Reply by thw0rted** [reply]
Is there a Prisma method argument that actually returns all related types?  I think I always have to write e.g. `foo.findMany({include: {bar: true, baz: true, ...}})`.  If I'm right and there isn't a way to pull back all relations automatically, when would an all-relations type be useful?

In case it helps, I've had a lot of success with the approach from https://github.com/prisma/prisma/discussions/10928#discussioncomment-8215402 .

**Reply by Destroy666x** [reply]
I don't think there is. There could also be. It'd be helpful for prototyping, for example. So that you don't have to add bunch of methods manually.

I'll take a look at that as manual possibility as at least it's simple enough, not sure it'll work with RxJS I'm also using though.

**Reply by Destroy666x** [reply]
I guess I could convert this https://github.com/prisma/prisma/blob/b268a402a4b137df7d9dfe567f2c12c03d84df74/packages/client/scripts/default-index.d.ts#L77 and the other type to work with Observables instead of Promises but ye...

**Reply by Destroy666x** [reply]
Ye I don't know about this. In theory it works with:
```
import { Observable } from 'rxjs';

// The type of an Observable
export declare type ObservableType<T extends Observable<any>> = T extends Observable<infer U> ? U : unknown;

// The type of Observable returned by function
export declare type ObservableReturnType<T extends (...args: any[]) => Observable<any>> = ObservableType<ReturnType<T>>;
```
\+
```
import { ObservableReturnType } from "@/shared/types";
import { OrdersRepository } from "./orders.repository";

export declare type FullOrder = ObservableReturnType<OrdersRepository['getFullOrder']>
```

But...
- it removes the possibility to use return type hint in function definition = less clarity
- since e.g. `findUnique` can return `null`, the automatic type adds `| null`. Using it in places where `Order` is 100% predictable (e.g. `OrderConfirmedEvent`) adds completely unnecessary overhead of checking for `null`
- and while having those disadvantages, it doesn't really remove the need of manual definitions as mentioned

So not a big fan @thw0rted Other manual solutions seem more comfortable as they eliminate problems 1 and 2, even though they need more boilerplate code. Unless I'm missing something.

**Reply by thw0rted** [reply]
1. When you say "use return type hint in function definition", if you mean you want explicit return types on all your functions -- this was my habit too, and I had to train myself to break it specifically for Prisma.  The fact is that the library put a ton of work into generating correct, detailed types, and you're only making your life harder if you try to explicitly define your own.  Always use inferred return types when passing out a Prisma result and you're going to be a lot happier.
2. Use `findUniqueOrThrow` and the return type does not include `null`.

**Comment by RadomirNowak** [comment]
For anyone still struggling - I wanted to be as typesafe as possible without having to repeat code, so I went with [this answer](https://github.com/prisma/prisma/discussions/10928#discussioncomment-8215402) and simply used the `satisfies` for the `include` part, like so:
```
const userInclude = {
  include: {
    clients: {
      select: {
        id: true,
        name: true,
      }
    }
  }
} satisfies Prisma.UserDefaultArgs;

const users = prisma.users.findUnique({
  where: {
    id: 'user-id'
  },
  include: userInclude.include,
})

const clients = users.clients; // <-- clients is now properly typed to { id: string; name: string }
```
